### PR TITLE
Handle tuple conversions and extension method `this` argument conversions in NullableWalker

### PR DESCRIPTION
**Pull request by cston**

### CODE REVIEW

**Comment by cston** [comment]
@dotnet-bot test this

**Comment by cston** [comment]
@dotnet/roslyn-compiler please review.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 1]
```diff
@@ -1858,6 +1864,7 @@ private Conversion ClassifyImplicitNullableConversion(TypeSymbol source, TypeSym
 
         private Conversion GetImplicitTupleLiteralConversion(BoundTupleLiteral source, TypeSymbol destination, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
         {
+            Debug.Assert(!IncludeNullability); // Top-level nullability is ignored.
```
[the remark this answers is not in the document]
Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: This method is not used with `IncludeNullability` currently, and the delegate below does not handle top-level nullability.

---
In reply to: [215710436](https://github.com/dotnet/roslyn/pull/29679#discussion_r215710436) [](ancestors = 215710436)

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 2]
```diff
@@ -1457,6 +1457,7 @@ public Conversion ClassifyImplicitExtensionMethodThisArgConversion(BoundExpressi
 
             if (sourceExpressionOpt?.Kind == BoundKind.TupleLiteral)
             {
+                Debug.Assert(!IncludeNullability); // Top-level nullability is ignored.
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: > This code path is not used when IncludeNullability is set. If it is in the future (when this assert fails), we'll need to update the delegate below.

Consider capturing this in the comment instead. Something like: "At the moment this code path is taken only under this condition, if that changes, the code below will need an adjustment". And for other similar places. 

---
In reply to: [215710774](https://github.com/dotnet/roslyn/pull/29679#discussion_r215710774) [](ancestors = 215710774,215709176)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 3]
```diff
@@ -32604,9 +32668,6 @@ static void F((B?, B) b)
 }";
             var comp = CreateCompilation(new[] { source, NonNullTypesTrue, NonNullTypesAttributesDefinition });
             comp.VerifyDiagnostics(
-                // (12,21): warning CS8604: Possible null reference argument for parameter 'a' in 'A.implicit operator C(A a)'.
```
[the remark this answers is not in the document]
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Logged [https://github.com/dotnet/roslyn/issues/29699](https://github.com/dotnet/roslyn/issues/29699). Thanks.

---
In reply to: [215729273](https://github.com/dotnet/roslyn/pull/29679#discussion_r215729273) [](ancestors = 215729273)

**Pull request merged**